[![Build Status](https://travis-ci.org/iqbal-lab-org/gramtools.svg?branch=master)](https://travis-ci.org/iqbal-lab-org/gramtools)

# gramtools
**TL;DR** Genome inference using prior information encoded as a reference graph.

Gramtools builds a population reference genome (PRG) from a set of variants. 
Given sequence data from an individual, the graph is annotated with coverage and genotyped. 
 
 A personalised reference genome for the sample is inferred and new variation discovered 
 against it. You can then build a new PRG from the initial + the new variants, and carry on forever!

## Install
```
pip install wheel git+https://github.com/iqbal-lab-org/gramtools
```
Requirements:
* Python >= 3.6
* pip >= 20.0.2
* C++17 compatible compiler: g++ >=8 (tested), clang >=7 (untested)
<br>

If `sudo` is unavailable, we recommend using a Python virtual enviroment:
```
python -m venv gram_ve && source gram_ve/bin/activate
pip install pip==20.0.2
pip install wheel git+https://github.com/iqbal-lab-org/gramtools
```

A singularity definition file (in containers/singularity) can also be used to build a gramtools image.

## Use

```
Gramtools

Usage: 
    gramtools [-h] [--debug] [--force] subcommand
    
    Subcommands:
        gramtools build -o GRAM_DIR --ref REFERENCE
                       (--vcf VCF [VCF ...] | --prg PRG)
                       [--kmer_size KMER_SIZE]

        gramtools genotype -i GRAM_DIR -o GENO_DIR
                          --reads READS [READS ...] --sample_id SAMPLE_ID
                          [--ploidy {haploid,diploid}]
                          [--max_threads MAX_THREADS] [--seed SEED]

        gramtools discover -i GENO_DIR -o DISCO_DIR
                          [--reads READS [READS ...]]

        gramtools simulate --prg PRG
                           [--max_num_paths MAX_NUM_PATHS]
                           [--sample_id SAMPLE_ID] [--output_dir OUTPUT_DIR]
```

### Subcommands explained
1) [build](https://github.com/iqbal-lab-org/gramtools/wiki/Commands%3A-build) - 
given a VCF and reference or a prg file, construct the graph.
    * `--kmer_size`: used for indexing the graph in preparation for
       `genotype`. higher `k` <=> faster `genotype`, but `build` output will consume more 
       disk space.

2) [genotype](https://github.com/iqbal-lab-org/gramtools/wiki/Commands%3A-genotype) - 
    map reads to a graph generated in `build` and genotype the graph. Produces genotype calls (VCF)
    and a personalised reference genome (fasta).
    * `--reads`: 1+ reads files in (fasta/fastq/sam/bam/cram) format
    * `--sample_id`: displayed in VCF & personalised reference outputs

3) [discover](https://github.com/iqbal-lab-org/gramtools/wiki/Commands%3A-discover) - 
discovers new variation against the personalised reference genome from `genotype` using
 one or more variant callers (currently: cortex).
 
4) simulate- samples paths through a prg, producing a fasta of the paths and a genotyped JSON
of the variant bubbles the path went through.
    * `--prg`: a prg file as output by `build`

Examples, documentation, and planned future enhancements can be found in the [wiki](https://github.com/iqbal-lab-org/gramtools/wiki).

## Reference documentation

A [doxygen](http://doxygen.nl/) formatted documentation can be generated by running 
```doxygen doc/Doxyfile.in```
from inside the gramtools directory.

The documentation gets generated in doc/html/index.html and provides a useful reference for all files, classes, functions and data structures in gramtools.

## License

MIT
